torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -604,126 +605,122 @@ void ONNXUpdateTypeFromTensor(
+    std::shared_ptr<Graph>& graph,
+    size_t outputs_index,
+    PyObject* output_obj,
+    bool onnx_shape_inference) {
+  auto index_check = [&]() {
+    TORCH_INTERNAL_ASSERT(
+        outputs_index >= 0 && outputs_index <= graph->outputs().size(),
+        "Incorrect number of elements provided as example outputs.");
+  };
+
+  index_check();
+
+  if (THPVariable_Check(output_obj)) {
+    at::Tensor var = reinterpret_cast<THPVariable*>(output_obj)->cdata;
+    ONNXUpdateTypeFromTensor(
+        graph->outputs().at(outputs_index), var, onnx_shape_inference);
+    outputs_index++;
+  } else if (PyTuple_Check(output_obj)) {
+    size_t tuple_len = PyTuple_GET_SIZE(output_obj);
+    for (size_t i = 0; i < tuple_len; ++i) {
+      outputs_index = ONNXAssignOutputShape(
+          graph,
+          outputs_index,
+          PyTuple_GET_ITEM(output_obj, i),
+          onnx_shape_inference);
+    }
+  } else if (PyList_Check(output_obj)) {
+    size_t list_len = PyList_GET_SIZE(output_obj);
+    if (HasSequenceTypeOutput(graph->outputs().at(outputs_index)->node())) {
+      if (list_len > 0) {
+        auto list_elem = PyList_GET_ITEM(output_obj, 0);
+        TORCH_INTERNAL_ASSERT(THPVariable_Check(list_elem));
+        auto& var = reinterpret_cast<THPVariable*>(list_elem)->cdata;
+        for (size_t i = 1; i < list_len; ++i) {
+          list_elem = PyList_GET_ITEM(output_obj, i);
+          TORCH_INTERNAL_ASSERT(THPVariable_Check(list_elem));
+          auto& new_var = reinterpret_cast<THPVariable*>(list_elem)->cdata;
+          TORCH_CHECK(
+              var.scalar_type() == new_var.scalar_type(),
+              "Unsupported sequence type in model outputs. ONNX supports sequences of elements of the same data type.");
+        }
+        auto elem_type = graph->outputs()
+                             .at(outputs_index)
+                             ->type()
+                             ->castRaw<ListType>()
+                             ->getElementType()
+                             ->cast<TensorType>();
+        elem_type = elem_type->withScalarType(var.scalar_type());
+        graph->outputs()
+            .at(outputs_index)
+            ->setType(MergeInferredType(
+                graph->outputs().at(outputs_index)->type(),
+                ListType::create(elem_type)));
+      } else {
+        graph->outputs()
+            .at(outputs_index)
+            ->setType(graph->outputs().at(outputs_index)->type());
+      }
+      outputs_index++;
+    } else {
+      // When torch output is a list type, but ONNX node is not a
+      // sequence type. Like prim::ListConstruct
+      for (size_t i = 0; i < list_len; ++i) {
+        outputs_index = ONNXAssignOutputShape(
+            graph,
+            outputs_index,
+            PyList_GET_ITEM(output_obj, i),
+            onnx_shape_inference);
+      }
+    }
+  } else if (PyDict_Check(output_obj)) {
+    // Support for dict data type is limited to fixed size dictionaries in
```

Comment:
Do we need a check on dictionary to see if it is of fixed size?

Reply:
We aren't able to check it. the obj here is the example output we get when exporting, and we assume the graph always produce dict output of that size. The warning of dict as input/output use case is emitted elsewhere.